### PR TITLE
[TT-9111] Fix failing smoke tests, add linter, update docs

### PR DESCRIPTION
**Pull request by titpetric**
https://tyktech.atlassian.net/browse/TT-9111

### CODE REVIEW

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit:  
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6494124745)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: c24e4a4ad9ac83eb4e268a822ca4816de24a3130 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6494142457)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: ea0482d188415c72da25aa2a8e0723a76a4a1134 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6494196540)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 5e3e92f43e5f1d5c164e5aae56e0767101751de2 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6494350619)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 2b78eaae894458de6d72462f343f9bbb5c029883 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6494864722)

File: ci/tests/plugin-compiler/test.sh [hunk 1]
```diff
@@ -1,4 +1,11 @@
 #!/bin/bash
+function setup {
+	# Setup required env vars for docker compose
+	export GATEWAY_IMAGE=${GATEWAY_IMAGE:-"tykio/tyk:${tag}"}
```
Comment (jeffy-mathew) [inline ci/tests/plugin-compiler/test.sh:4]: this is `tyk-gateway` right?

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 6964628c7155020512d513617b623413232049d6 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6494950708)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 0edb026a8dc450e103d52014516097d3754ed216 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6495251399)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 4531f2bf935e19c0d22a83055b39c14b719e2949 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6495249051)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: f97080b9b9faad827c4e4c6b907321f8c1a0a5d4 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6495311436)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: f3d3ed0f9a5c4e5b2d9f1fe7714fbe8517681fcf 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6495997492)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 63a01483843f685874c7fc2eca76381beccecf44 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6496217636)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: ac3d88f0583597dd0118432ae7653944c87dd3a0 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6496409977)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 1de47c4bf0bab68137238395680d6f684c1598b6 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6496605205)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 9d0be7a827cc9ca6dc497ca68bba99372a7951fa 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6496871810)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 3de18cecae373794ad0d83784e3889b858f12761 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6497070749)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: ec04d731ee1e2df53e6f4beab204fd15d547755d 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6497090767)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 7ce265cde9e883371fba06d9cb4f2d574838ce62 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6505501165)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 5b8b283c2f845b2f4498aaded90dbbb8df4ad2de 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6505496895)

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5609/merge
Commit: 133e09f1fae63f7d759b44a418f33f6a808f7be5 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6505456467)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 01d1c5a8af10fec51727c38ce3b8d4c2760fb050 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6506026352)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 8f1538cbb73ac946b809c9c721be37f76eddc056 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6506433868)

**Review by jeffy-mathew — APPROVED** [review]

**Comment by buger** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5609/merge
Commit: ee37824c07ae49986f6649a555c86c6df486e2e4 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6506712742)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: e876653b553cb1b014bcbde9d44b9cf9817b369e 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6507066404)

**Review by buger — APPROVED** [review]
Look safe!

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -393,29 +393,37 @@ jobs:
...
+          role-to-assume: arn:aws:iam::754489498669:role/ecr_rw_tyk
+          role-session-name: cipush
+          aws-region: eu-central-1
+          # Don't mask to pass it across job boundaries
+          mask-aws-account-id: false
+
+      - uses: aws-actions/amazon-ecr-login@v1
+        id: ecr
+
+      - name: "Run ci/tests"
```
Comment (alephnull) [inline .github/workflows/release.yml:419]: These tests are outside the templates and perhaps we could remove this altogether.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -425,6 +433,24 @@ jobs:
...
+      - name: "Run smoke-tests"
+
+        # This job only runs whenever a tag is created. A tag is required
+        # for a functional plugin compiler build for when the GO_GET=1 env
+        # is provided. The plugin compiler cannot fetch the referenced
+        # commit from a PR, but requires a /heads or /tags reference.
+        #
+        # See https://github.com/golang/go/issues/31191 for more info.
+
+        if: startsWith(github.ref, 'refs/tags')
```
Comment (alephnull) [inline .github/workflows/release.yml:446]: Now that we build images and push them in PRs, we should certainly run these in the PR context.

File: ci/tests/api-functionality/test.sh
Comment (alephnull) [inline ci/tests/api-functionality/test.sh:1]: This test is also in the upgrade tests afaict?

File: .github/workflows/release.yml (hunk 3 above)
Comment (titpetric) [reply .github/workflows/release.yml:446]: We can't due to the caveat listed above (no way to `go get` from a PR commit, it needs to be in heads or tags)

File: .github/workflows/plugin-compiler-build.yml [hunk 4]
```diff
@@ -45,7 +45,7 @@ jobs:
             type=semver,pattern=v{{version}}
             type=semver,pattern=v{{major}}.{{minor}}
             type=semver,pattern={{raw}}
-            type=raw,value=${{ github.sha }},enable=${{ !startsWith(github.ref, 'refs/tags') }}
+            type=raw,value=${{ github.sha }}
```
[the remark this answers is not in the document]
Comment (ermirizio) [reply .github/workflows/plugin-compiler-build.yml:48]:
```suggestion
           type=sha,format=long
```

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -393,29 +393,37 @@ jobs:
...
+      - uses: aws-actions/configure-aws-credentials@v2
+        with:
+          role-to-assume: arn:aws:iam::754489498669:role/ecr_rw_tyk
+          role-session-name: cipush
+          aws-region: eu-central-1
+          # Don't mask to pass it across job boundaries
+          mask-aws-account-id: false
+
+      - uses: aws-actions/amazon-ecr-login@v1
+        id: ecr
```
Comment (ermirizio) [inline .github/workflows/release.yml:417]:
```suggestion
      - uses: aws-actions/amazon-ecr-login@v1
        id: ecr
        with:
          mask-password: 'true'
```

**Review by ermirizio — CHANGES_REQUESTED** [review]

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: c3bc27072135c9b706304698b28cc8e2d60a3b9c 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6546528575)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: f2385bc4e331905e160ffbd0dbfbd2dbe255d5a5 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6546529407)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: f6f6b9b31ceb07f0b9222755ef8b0fc860ef93f6 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6547329395)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: ccf2d428d9a890fb9d8efe6e5548d83ec09dfa5b 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6547378077)

**Review by ermirizio — APPROVED** [review]

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: 1a9efbe77b30afacf450f4c2a0d887b1314911f3 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6547935799)

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5609/merge
Commit: d3832062810e804469f75d2904eb77018b3dc18b 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6547959002)

**Pull request merged**